### PR TITLE
[MM-11315] Implement loadtest for channel creation/deletion

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

This PR adds channel creation/deletion to loadtests.

#### Ticket

https://mattermost.atlassian.net/browse/MM-11315

### CODE REVIEW

File: loadtest/run.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 func RunTest(test *TestRun) error {
 	r := rand.New(rand.NewSource(0))
 
-	interruptChannel := make(chan os.Signal)
+	interruptChannel := make(chan os.Signal, 1)
```
Comment (lieut-data) [inline loadtest/run.go:33]: Can you explain why this change is here?

File: loadtest/tests.go [hunk 2]
```diff
@@ -290,6 +291,135 @@ func createPost(c *EntityConfig, team *UserTeamImportData, channelId string) {
...
+	}
+
+	if choice.Item == PUBLIC_CHANNEL || choice.Item == PRIVATE_CHANNEL {
+		channel := &model.Channel{}
+		channel.TeamId = teamId
+		channel.Name = makeChannelName(c.r.Int())
+		channel.DisplayName = makeChannelDisplayName(c.r.Int())
+
+		if choice.Item == PUBLIC_CHANNEL {
+			channel.Type = "O"
```
Comment (lieut-data) [inline loadtest/tests.go]: Can we use the `model.*` constants for these instead?

File: loadtest/tests.go [hunk 3]
```diff
@@ -290,6 +291,135 @@ func createPost(c *EntityConfig, team *UserTeamImportData, channelId string) {
...
+
+		newChannel, resp := c.Client.CreateChannel(channel)
+		if resp.Error != nil {
+			return nil, resp.Error
+		}
+
+		return newChannel, nil
+	} else if choice.Item == DIRECT_CHANNEL || choice.Item == GROUP_CHANNEL {
+
+		users, resp := c.Client.GetUsers(0, 100, "")
```
Comment (lieut-data) [inline loadtest/tests.go]: If I follow this correctly, this will limit us to a maximum of 10,000 direct message channels. In practice, DMs account for 95+% of the channels in a production table once an installation hits 20,000+ users.

This also incurs an extra "GetUsers" cost that isn't technically part of the action of creating a new channel, and thus doesn't simulate a user entity very well.

Could we leverage our pre-existing knowledge of the users on the system to drive this instead?

**Review by lieut-data — CHANGES_REQUESTED** [review]

File: loadtest/tests.go (hunk 3 above)
Comment (streamer45) [reply loadtest/tests.go]: Yes, you are right. This is not the first time that I encounter this issue and it's going to be addressed in a brief review document I am working on (and will publish soon), essentially expressing my thoughts as a completely new user/developer approaching this codebase.

Coming back to this PR. Is having all the users loaded in memory before starting the tests and randomly picking some the best way to simulate this case?  

I am thinking that maybe a `/users/search` call might be more close to reality. That obviously poses different problems as the user needs to know what to look for.  
Maybe the two approaches could be used in conjunction. We pick a user randomly from the in memory state and then we have the user entity search for him.

These are just some thoughts, please let me know what you think.

File: loadtest/tests.go (hunk 2 above)
Comment (streamer45) [reply loadtest/tests.go]: Yep

File: loadtest/run.go (hunk 1 above)
Comment (streamer45) [reply loadtest/run.go:33]: This is unrelated to the PR. I ran into a warning that `signal.Notify()` could miss some signals if the channel was not buffered. I read the relevant part in the [godoc](https://golang.org/pkg/os/signal/#Notify) and it was suggested to use a buffered channel.

Comment (lieut-data) [reply loadtest/run.go:33]: Makes sense! Learned something new :)

File: loadtest/tests.go (hunk 3 above)
Comment (lieut-data) [reply loadtest/tests.go]: Agreed with both your ideas: have a pre-existing knowledge of users, and then drive a few queries to `/users/search` like a human would when constructing such a channel.

File: loadtestconfig.default.json [hunk 4]
```diff
@@ -75,7 +75,11 @@
         "UserProfileUpdateNicknameChance": 0.01,
         "UserProfileUpdatePositionChance": 0.107,
         "UserProfileUpdateEmailChance": 0.003,
-        "UserProfileUpdateImageChance": 0.865
+        "UserProfileUpdateImageChance": 0.865,
+        "PublicChannelCreationChance": 0.12,
```
Comment (crspeller) [inline loadtestconfig.default.json:81]: Where are these numbers from?

File: loadtest/tests.go [hunk 5]
```diff
@@ -290,6 +291,147 @@ func createPost(c *EntityConfig, team *UserTeamImportData, channelId string) {
...
+			return nil, resp.Error
+		}
+
+		return newChannel, nil
+	}
+
+	return nil, nil
+}
+
+func actionCreateDeleteChannel(c *EntityConfig) {
```
Comment (crspeller) [inline loadtest/tests.go:406]: Why combined? Is it a realistic simulation to create a channel then immediately delete it?

**Review by crspeller — CHANGES_REQUESTED** [review]

File: loadtest/tests.go (hunk 5 above)
Comment (streamer45) [reply loadtest/tests.go:406]: I was kind of following a pattern already used in other tests. The way it's developed I find it more of a method to stress test this specific case rather than simulate anything. 

That said, your concern is a valid one and I think it might require a more involved discussion over the current state of this tool and the direction we'd like it to go.

I've recently started this process by posting my review of `loadtest` in [Developers: Performance](https://community.mattermost.com/core/pl/xd5httxcbtyoi8xn6jkjnd8a8w). It would be great to hear your thoughts too.

File: loadtestconfig.default.json (hunk 4 above)
Comment (streamer45) [reply loadtestconfig.default.json:81]: The `UserProfile` part comes from some analytics data I was given specific to this case.
The values in the `ChannelCreation` section were calculated from the default config itself using `NumChannelsPerTeam` and other related values.  In this case I didn't have access to any relevant statistics.

**Review by lieut-data — APPROVED** [review]

**Review by crspeller — APPROVED** [review]
